### ISSUE TITLE
Expose lower-level constructs

### ISSUE THREAD

**Issue by SamVerschueren** [isomorphic-git/isomorphic-git#1495]
Hi everyone 👋 !

I was wondering if it would make sense to expose lower-level constructs like the managers and the models. That would give users of `isomorphic-git` more flexibility in building features that are currently not supported by `isomorphic-git`. As an example, the issue I just created, https://github.com/isomorphic-git/isomorphic-git/issues/1494, could probably be built myself if I had access to the `GitIndexManager`.

After that I could always contribute back if we agree on an API that might be helpful for others to use as well. But `isomorphic-git` has such good low-level building blocks, and it's sometimes annoying that it cannot be used directly.

**Comment by mojavelinux** [comment]
We have to be very cautious when exposing internal/utility object-oriented APIs because it increases the API surface area of the library. That can make isomorphic-git harder to refactor in the future without ripping effects. I would be in favor of adding more plumbing functions like we have with resolveRef. But to expose APIs like GitRefManager moves isomorphic-git away from being a functional library and I'm not in favor of it.

**Comment by SamVerschueren** [comment]
Doing a refactor could also be released as a semver major because it's breaking. However, I do see what you are pointing out here which definitely makes sense. I'm trying to find a solution for myself, and maybe also others, to enable more flexibility without constantly taking time from maintainers like you and @jcubic who do this in their spare time.

**Comment by jcubic** [comment]
I personally think that having more objects like `GitRefManager` is not that bad. They can be used for specific low-level handling of the index, cache, etc.

But it's also a good idea to make this a separate project. And later if it will be needed it can be integrated. We can even make it in isomorphic-git org as an experiment.

Isomorphic-git is created using ES Modules so there is no reason not to dig deep ad import internal functions to implement this new library. If something is not exported (internally) we can modify it and export it so it can be used in this new project.

@SamVerschueren you can create a proof of concept of the new low-level API. We need to be very careful about what the API will look like.

I'm thinking about maybe `GitIndex` object that has stuff that is related to git index. And it doesn't have to have all the features. We can add them one by one when needed. When `GitIndex` has everything that is needed, we can integrate it with isomorphic-git and make it a major update.

**Comment by mojavelinux** [comment]
If we are to move away from only exposing a pure functional API, then I think it must be done as a separate project. Otherwise, that would be a huge change in what isomorphic-git is. However, it's perfectly reasonable for isomorphic-git itself to depend on this new library, effectively hiding it underneath it's current functional API. I don't think that we should be adding it directly to this project. In fact, I'd like to see those internal exports locked down, which Node.js / ES modules now allows. (See https://nodejs.org/api/packages.html#packages_exports).

**Comment by SamVerschueren** [comment]
Do you mean it should be something like `@isomorphic-git/core` which has all the managers, models and utilities? That module is then used by `isomorphic-git` and this library just contains the public API? If you then want low-level constructs, you could import `@isomorphic-git/core`, otherwise you could use the public API surface of `isomorphic-git`.

**Comment by mojavelinux** [comment]
That seems like a reasonable name to me. That could expand to other subprojects if there are clear lines to draw between them. For example, we could consider @isomorphic-git/managers and @ismorphic-git/models, matching the folder names currently used.

**Comment by jcubic** [comment]
Maybe in the future, it can be `@isomorphic-git/core` right now I was thinking about an experiment like `@isomorphic-git/git-index`, this will be a PoC to see how it will work.

**Comment by mojavelinux** [comment]
I agree that if we go too fine-grained, it becomes a project maintenance and infrastructure nightmare. I think logical groupings makes the most sense. The Babel project does it this way. Those packages can even be put in the same monorepo. (Though, one could argue that using a monorepo allows you to get more fine-grained).

**Comment by SamVerschueren** [comment]
In order to proceed here, as I can start working on this today, what is the path forward here?

Do I just create an `@isomorphic-git/git-index` package and copy over everything needed for it to work? Do we look into some kind of Nx/Lerna monorepo architecture?

**Comment by jcubic** [comment]
I'm not sure. I can create an empty repo in the organization so you can fork it, but I'm not sure about NPM, we probably can't publish it that way because of permissions. Right now publishing is in CI/CD pipeline. BTW @mojavelinux I don't know if we can migrate to GitHub Actions without access to NPM.

Maybe we should ask @wmhilton what he thinks about this. He is the original author. But I'm not sure if he will reply, he is not as active as at the beginning.

**Comment by jcubic** [comment]
I've pinged him on Twitter.

**Comment by mojavelinux** [comment]
>  what is the path forward here

I'm not sure we've agreed yet on how to organize the objects into packages. I'd like to see a more thorough proposal that outlines where we are going with this if we decide to pursue it. I want to avoid the situation where one package is created just to access a single method, and then have the idea abandoned. If these objects are going to pulled out into separate packages, we need to know how far it is going to go and how they will be incorporated into the isomorphic-git build. As such, it might be best to put together a proof of concept in your own namespace to demonstrate how it will work. I don't think we're ready to make any commitments in isomorphic-git yet.

As I said earlier, exposing an object-oriented interface is a big deal for isomorphic-git. William was very careful to design a functional API. This change potentially takes us in a very different direction, even if they are offered as auxiliary packages. We need to understand how this is going to fit into the isomorphic-git story and make sure we aren't losing the focus of the design in the process.

**Comment by mojavelinux** [comment]
> I can create an empty repo in the organization so you can fork it

I would much rather see a proof of concept developed in a personal namespace first. As soon as we have a repository in this organization, we have to maintain it. And we haven't even decided if we are going to do this yet. So I don't think that's a good idea.

**Comment by jcubic** [comment]
Yes, I was about to write about personal repo but didn't add that comment. I agree it's better if PoC is created on your own namespace.

Also if the code is short and uses the internals of Isomorphic-git it can be just part of the FAQ. Also, we can make some internals documented and make them official, but we will need to be careful before creating public documentation. Maybe rethink how the code is organized.

**Comment by wmhilton** [comment]
> BTW @mojavelinux I don't know if we can migrate to GitHub Actions without access to NPM.

Just wanted to pop in and say I'm happy to add you two to the isomorphic-git npm team. ❤️

As far as a low-level API... I'll just say that the public API I am quite proud of. The internal API with its Managers and Models... well I'm not "ashamed" of it per se but it's not a low-level API that I enjoy using.

I have seen projects go too far in the hyper-modular direction, and consciously chose to design isomorphic-git to be batteries-included, but not too scary if you wanted to fork it and tinker with the batteries.

For what it's worth, I like the idea of a high-level `updateIndex` function. ❤️

**Comment by wmhilton** [comment]
Thank you @SamVerschueren for helping make isomorphic-git even better

**Comment by wmhilton** [comment]
Thank you @jcubic and @mojavelinux for being such good stewards of the code and the community!
